### PR TITLE
[WIP] *: add etcd backup controller logidc

### PR DESCRIPTION
**Pull request by fanminshi**

### CODE REVIEW

**Comment by fanminshi** [comment]
test this code manually.

Create Operator and EtcdCluster

Operator Deployement:
```yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: etcd-operator
spec:
  replicas: 1
  template:
    metadata:
      labels:
        name: etcd-operator
    spec:
      containers:
      - name: etcd-operator
        image: quay.io/coreos/etcd-operator:v0.6.0
        env:
        - name: MY_POD_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        - name: MY_POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
```

EtcdCluster Deployement

```yaml
apiVersion: "etcd.database.coreos.com/v1beta2"
kind: "EtcdCluster"
metadata:
  name: "example-etcd-cluster"
spec:
  size: 3
  version: "3.1.8"
```

Create CRD:
`kubectl create -f example/etcd-backup-operator/crd.yaml`

```yaml
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: etcdbackups.etcd.database.coreos.com
spec:
  group: etcd.database.coreos.com
  version: v1beta2
  Scope: Namespaced
  names: 
    kind: EtcdBackup
    plural: etcdbackups
```

Create AWS secret:
`kubectl create -f example/etcd-backup-operator/secret.yaml`

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: aws
type: Opaque
data:
  region: <your-region-base64>
  aws_access_key_id: <your-acces-id-base64>
  aws_secret_access_key: <your-access-key-base64>
```


kubectl create -f example/etcd-backup-operator/deploy.yaml

```yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: etcd-backup-operator
spec:
  replicas: 1
  template:
    metadata:
      labels:
        name: etcd-backup-operator
    spec:
      containers:
      - name: etcd-backup-operator
        image: gcr.io/coreos-k8s-scale-testing/etcd-backup-operator:fanmin
        imagePullPolicy: Always
        env:
        - name: MY_POD_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        - name: MY_POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
```

create backup under enkins-etcd-operator/prefix/***

`kubectl create -f example/etcd-backup-operator/backup.yaml`

```yaml
apiVersion: "etcd.database.coreos.com/v1beta2"
kind: "EtcdBackup"
metadata:
  name: example-etcd-cluster
spec:
  clusterName: example-etcd-cluster
  storageType: s3
  s3:
    s3Bucket: jenkins-etcd-operator
    prefix: prefix
    awsSecret: aws
```
check etcd-backup-operator logs
```sh
kubectl logs -f etcd-backup-operator-1240628667-73lmf
time="2017-09-29T21:48:28Z" level=info msg="Go Version: go1.9"
time="2017-09-29T21:48:28Z" level=info msg="Go OS/Arch: linux/amd64"
time="2017-09-29T21:48:28Z" level=info msg="etcd-backup-operator Version: 0.0.1+git"
time="2017-09-29T21:48:28Z" level=info msg="Git SHA: eacd485"
time="2017-09-29T21:48:45Z" level=info msg="Event(v1.ObjectReference{Kind:"Endpoints", Namespace:"default", Name:"etcd-backup-operator", UID:"5c06e54d-a55e-11e7-aeb0-080027059fff", APIVersion:"v1", ResourceVersion:"114614", FieldPath:""}): type: 'Normal' reason: 'LeaderElection' etcd-backup-operator-1240628667-73lmf became leader"
time="2017-09-29T21:48:45Z" level=info msg="starting backup controller"
time="2017-09-29T21:50:22Z" level=info msg="getMaxRev: member example-etcd-cluster-0000 revision (1)"
time="2017-09-29T21:50:22Z" level=info msg="getMaxRev: member example-etcd-cluster-0001 revision (1)"
time="2017-09-29T21:50:22Z" level=info msg="getMaxRev: member example-etcd-cluster-0002 revision (1)"
2017-09-29 21:50:22.295420 I | saving backup for cluster (example-etcd-cluster)
```

show backup
```sh
aws s3 ls jenkins-etcd-operator/prefix/v1/default/example-etcd-cluster/
2017-09-29 14:50:23      24608 3.1.8_0000000000000001_etcd.backup
```

**Comment by fanminshi** [comment]
cc/ @hongchaodeng  please take a look of general framework. it currently compiles and can do etcd cluster backup.

File: pkg/etcd-backup-controller/controller.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
+package controller
```
Comment (hongchaodeng) [inline pkg/etcd-backup-controller/controller.go]: dir should be `etcd-backup-operator/controller`

File: cmd/backup-operator/main.go [hunk 2]
```diff
@@ -74,5 +76,9 @@ func createRecorder(kubecli kubernetes.Interface, name, namespace string) record
 }
 
 func run(stop <-chan struct{}) {
-	// TODO
+	c := controller.New()
+	err := c.Start(context.TODO())
+	if err != nil {
+		logrus.Infof("operator stopped with %v", err)
```
Comment (hongchaodeng) [inline cmd/backup-operator/main.go]: What was the case that it would return error again?

File: hack/build/etcd-backup-operator/push [hunk 3]
```diff
@@ -5,10 +5,10 @@ if ! which docker > /dev/null; then
 	exit 1
 fi
 
-: ${IMAGE:?"Need to set IMAGE, e.g. gcr.io/coreos-k8s-scale-testing/etcd-backup-operator"}
+: ${IMAGE:?"Need to set IMAGE, e.g. gcr.io/coreos-k8s-scale-testing/etcd-backup-operator:fanmin"}
```
Comment (hongchaodeng) [inline hack/build/etcd-backup-operator/push]: revert

File: hack/build/etcd-backup-operator/push [hunk 4]
```diff
@@ -5,10 +5,10 @@ if ! which docker > /dev/null; then
...
 
 echo "building container..."
-docker build --tag "${IMAGE}" -f hack/Dockerfile . 1>/dev/null
+docker build --tag "${IMAGE}" -f hack/build/etcd-backup-operator/Dockerfile . 1>/dev/null
```
Comment (hongchaodeng) [inline hack/build/etcd-backup-operator/push]: separate this into another PR as fix

File: pkg/etcd-backup-controller/controller.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	key, err := cache.MetaNamespaceKeyFunc(newObj)
+	if err != nil {
+		panic(err)
+	}
+	b.queue.Add(key)
+}
+
+func (b *Backup) onDelete(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err == nil {
```
Comment (hongchaodeng) [inline pkg/etcd-backup-controller/controller.go]: What if err != nil ?

File: hack/build/etcd-backup-operator/push (hunk 4 above)
Comment (fanminshi) [reply hack/build/etcd-backup-operator/push]: kk

File: cmd/backup-operator/main.go (hunk 2 above)
Comment (fanminshi) [reply cmd/backup-operator/main.go]: the init() function is not yet implemented

```go
func (b *Backup) init(ctx context.Context) error {
	// TODO: check if crd exists
	return nil
}
```

it can return errors.

File: pkg/etcd-backup-controller/controller.go (hunk 1 above)
Comment (fanminshi) [reply pkg/etcd-backup-controller/controller.go]: do you mean `pkg/etcd-backup-operator/controller`?

Comment (hongchaodeng) [reply pkg/etcd-backup-controller/controller.go]: right

File: pkg/etcd-backup-controller/controller.go (hunk 5 above)
Comment (fanminshi) [reply pkg/etcd-backup-controller/controller.go]: maybe panic?

File: hack/build/etcd-backup-operator/push (hunk 4 above)
Comment (fanminshi) [reply hack/build/etcd-backup-operator/push]: done!

File: hack/build/etcd-backup-operator/push (hunk 3 above)
Comment (fanminshi) [reply hack/build/etcd-backup-operator/push]: done!

File: pkg/etcd-backup-controller/controller.go (hunk 1 above)
Comment (fanminshi) [reply pkg/etcd-backup-controller/controller.go]: done!

**Comment by fanminshi** [comment]
this pr is broken into https://github.com/coreos/etcd-operator/pull/1463 and https://github.com/coreos/etcd-operator/pull/1475. close in favor of those.
